systemtests/src/main/java/io/enmasse/systemtest/messaginginfra/MessagingInfraResourceManager.java

```diff
@@ -0,0 +1,228 @@
+        LOGGER.info("Going to clear all method resources");
+        while (!methodResources.empty()) {
+            methodResources.pop().run();
+        }
+        methodResources.clear();
+        pointerResources = classResources;
+    }
+
+    //------------------------------------------------------------------------------------------------
+    // Client factories
+    //------------------------------------------------------------------------------------------------
+
+    public void initFactories(AddressSpace addressSpace) {
+        amqpClientFactory = new AmqpClientFactory(addressSpace, environment.getSharedDefaultCredentials());
+        mqttClientFactory = new MqttClientFactory(addressSpace, environment.getSharedDefaultCredentials());
+    }
+
+    public void initFactories(AddressSpace addressSpace, UserCredentials cred) {
+        amqpClientFactory = new AmqpClientFactory(addressSpace, cred);
+        mqttClientFactory = new MqttClientFactory(addressSpace, cred);
+    }
+
+    public void initFactories(IoTProject project, UserCredentials credentials) {
+        String addSpaceName = project.getSpec().getDownstreamStrategy().getManagedStrategy().getAddressSpace().getName();
+        initFactories(kubeClient.getAddressSpaceClient(project.getMetadata()
+                .getNamespace()).withName(addSpaceName).get(), credentials);
+    }
+
+    public void closeAmqpFactory() throws Exception {
+        if (amqpClientFactory != null) {
+            amqpClientFactory.close();
+            amqpClientFactory = null;
+        }
+    }
+
+    public void closeMqttFactory() {
+        if (mqttClientFactory != null) {
+            mqttClientFactory.close();
+            mqttClientFactory = null;
+        }
+    }
+
+    public void closeFactories() throws Exception {
+        closeAmqpFactory();
+        closeMqttFactory();
+    }
+
+    public AmqpClientFactory getAmqpClientFactory() {
+        return amqpClientFactory;
+    }
+
+    public MqttClientFactory getMqttClientFactory() {
+        return mqttClientFactory;
+    }
+
+    //------------------------------------------------------------------------------------------------
+    // Common resource methods
+    //------------------------------------------------------------------------------------------------
+
+    public <T extends HasMetadata> void createResource(T... resources) throws Exception {
+        for (T resource : resources) {
+            LOGGER.info("Create/Update of {} {} in namespace {}",
+                    resource.getKind(), resource.getMetadata().getName(), resource.getMetadata().getNamespace() == null ? "(not set)" : resource.getMetadata().getNamespace());
+            switch (resource.getKind()) {
+                case "MessagingInfra":
+                    createMessagingInfra((MessagingInfra) resource);
+                    waithForInfraUp((MessagingInfra) resource);
+                    scheduleDeletion(MessagingInfraCrd.getClient(), (MessagingInfra) resource);
```

Comment:
Should this be called before waitForInfraUp in case it fails and we want to clean up?